### DISCUSSION TITLE
Invalid login

### DISCUSSION THREAD

**Discussion by jpsn21** [jasonacox/Powerwall-Dashboard#377] [General]
Hi, Im getting this error when running pypowerwall. I have entered my details and confirmed they are correct. Any suggestions to fix?

pypowerwall  | Traceback (most recent call last):
pypowerwall  |   File "/usr/local/lib/python3.10/site-packages/pypowerwall/__init__.py", line 155, in _get_session
pypowerwall  |     self.auth = {'AuthCookie': r.cookies['AuthCookie'], 'UserRecord': r.cookies['UserRecord']}
pypowerwall  |   File "/usr/local/lib/python3.10/site-packages/requests/cookies.py", line 334, in __getitem__
pypowerwall  |     return self._find_no_duplicates(name)
pypowerwall  |   File "/usr/local/lib/python3.10/site-packages/requests/cookies.py", line 413, in _find_no_duplicates
pypowerwall  |     raise KeyError(f"name={name!r}, domain={domain!r}, path={path!r}")
pypowerwall  | KeyError: "name='AuthCookie', domain=None, path=None"
pypowerwall  | 
pypowerwall  | During handling of the above exception, another exception occurred:
pypowerwall  | 
pypowerwall  | Traceback (most recent call last):
pypowerwall  |   File "/app/server.py", line 109, in <module>
pypowerwall  |     pw = pypowerwall.Powerwall(host,password,email,timezone,cache_expire,timeout,pool_maxsize)
pypowerwall  |   File "/usr/local/lib/python3.10/site-packages/pypowerwall/__init__.py", line 138, in __init__
pypowerwall  |     self._get_session()
pypowerwall  |   File "/usr/local/lib/python3.10/site-packages/pypowerwall/__init__.py", line 164, in _get_session
pypowerwall  |     raise LoginError("Invalid Powerwall Login")
pypowerwall  | pypowerwall.LoginError: Invalid Powerwall Login
pypowerwall exited with code 0

Thanks,

**Comment by jasonacox** [comment]
The Powerwall is rejecting the login attempt and not sending back the token (cookie) to authorize access.

Can you try this:

```bash
# Install pypowerwall python package
pip install pypowerwall
python -m pypowerwall scan
```

Verify that you see your Powerwall on your network and note the IP address.  Try to browse to that network address and use the Customer login type and use the password to see if you can log in.

<img width="450" alt="image" src="https://github.com/jasonacox/Powerwall-Dashboard/assets/836718/cf87e727-f36a-4012-864f-45a478cc012d">

If that works, now try to connect manually by running python:

```bash
python3
```

```python
import pypowerwall
host = "x.x.x.x" # replace with address seen in scan above
password = "password" # replace with your customer password set on gateway
pw = pypowerwall.Powerwall(host,password)
pw.grid()
```

If that works, make sure the same settings are in the `pypowerwall.env` file:

```bash
cat pypowerwall.env
```

**Reply by jpsn21** [reply]
Thanks for the reply, 

When I run the powerwall scan it comes back with 192.168.0/24 which is not the address. My powerwall is on 192.168.1.14 and I can connect to that through the browser but it just comes up with the tesla pros app. When i put in the other address I get a can not connect error 

The login settings are correct but I am still getting invalid login error

**Reply by jasonacox** [reply]
> When I run the powerwall scan it comes back with 192.168.0/24

That's odd - I suspect you are missing a number.  It looks like the network address. If your Powerwall IP is 192.168.1.14, then the network would likely be 192.168.1.0/24.  If you enter that as your network, the scan should find your Powerwall. Regardless, the "Tesla Pros" app is your problem.

[EDIT] - BEFORE you address the Tesla Pros app, stop pypowerwall.  Since it keeps trying to log in (restart), it is possible that the Powerwall gateway is rejecting it due to high cpu load.

```bash
docker stop pypowerwall
```

>  I can connect to that through the browser but it just comes up with the tesla pros app

I had the same thing. I had to click the "continue" button at the bottom:

<img width="640" alt="image" src="https://github.com/jasonacox/Powerwall-Dashboard/assets/836718/afda61f1-6f82-4f03-840f-e93038799fca">

And then the "Upgrade Later" text on the next screen.

<img width="496" alt="image" src="https://github.com/jasonacox/Powerwall-Dashboard/assets/836718/0988ac1e-8603-4ebc-9761-b13f6134e5b2">

After doing this, it didn't ask again.  Here is the other discussion thread on that: https://github.com/jasonacox/Powerwall-Dashboard/discussions/367#discussioncomment-7278760

Restart pypowerwall:

```bash
docker start pypowerwall
```

**Comment by youzer-name** [comment]
I tried spinning up a Grafana container on another Raspberry Pi (not my Powerwall Dashboard host) to play around with the latest version of Grafana. I was able to import my dashboard, but I can't get past the Teslapro / login screen in the Powerwall animation panel.

This isn't a supported setup, so I don't expect anyone to spend too much time trying to figure it out, but I'm wondering if I'm just missing something simple that would make it work.

The setup is:
Raspberry Pi 1: hosting pw dashboard including pypowerwall
Raspberry Pi 2: hosting new grafana instance
Windows PC: accessing dashboard in a browser

After changing the panel JSON to use the IP of the host that is running pypowerwall, rather than 'window.location.hostname', I get the Tesla logon screen in place of the animation. The firmware version displays normally. I tried clicking through 'Continue' and 'Upgrade Later' within the Grafana panel, but that just takes me to the logon screen. I get an error if I try to login from inside the Grafana panel, and when I reload the page, I'm back to the first logon screen with the 'Continue' button.

I tried going all the way through the logon process from the GUI of the RPi that is hosting the new Grafana instance. That didn't help.

I tried stopping and restarting pypowerwall, 

I tried a different browser and a hard reload. 

I can successfully load the animation directly in the browser (http://pw_hostip:8675), but when Grafana tries to load it in the panel, I get the Tesla login screen. Is this just a limitation of the way the proxy works where I won't be able to get past the upgrade prompts with Grafana on a different host from the proxy?

EDIT: Plugging away at this a little more, I installed pypowerwall on the 2nd RPi and pointed the iframe to that URL. It works. I'm assuming the reason it won't work through the original pypowerwall has something to do with passing the login authentication from one machine (running Grafana) to the other (running pypowerwall). 

Is it ok to have two different pypowerwall containers pointed at the same Gateway 2 or am I asking for trouble doing that?

**Reply by jasonacox** [reply]
I suspect the mixed sources in the iFrame could be causing some of the issue (with auth). You could try running a pypowerwall instances on your 2nd RPi to just proxy the animation.  That way it would have the same origin as the dashboard.
